trees/trees.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package trees contains utility method for retrieving trees and acquiring objects (hashers,
+// signers) associated with them.
+package trees
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/google/trillian"
+	tcrypto "github.com/google/trillian/crypto"
+	"github.com/google/trillian/crypto/keys"
+	"github.com/google/trillian/crypto/sigpb"
+	"github.com/google/trillian/errors"
+	"github.com/google/trillian/merkle"
+	"github.com/google/trillian/merkle/rfc6962"
+	"github.com/google/trillian/storage"
+	"golang.org/x/net/context"
+)
+
+type treeKey struct{}
+
+// NewContext returns a ctx with the given tree.
```

Comment:
Can/should the existing `util.NewMapContext` / `util.NewLogContext` methods be replaced with this?

(If not, we should update them so the key type there is not `int`, as (now) [recommended](https://golang.org/pkg/context/#WithValue) for `WithValue`)